smtp-server/src/lib.rs

```diff
@@ -127,21 +144,40 @@ pub trait Config: Send + Sync {
         }
     }
 
+    /// `handle_mail` is an async function that returns either a single decision
+    /// in the case of the SMTP protocol, or an async stream of decisions in the
+    /// case of the LMTP protocol.
+    ///
+    /// For LMTP: there must be one such decision for each RCPT TO command that
+    /// succeeded (i.e. for each accepted recipient), which allows to indicate
+    /// that the message could be stored in the mailbox of some users, but not
+    /// in that of other users. This can happen for instance if their mail
+    /// quota is used up. The lifetimes of the borrow on the mail's data stream
+    /// is limited to the duration of the async call; this reference may not be
+    /// retained by the returned stream. The async function must consume the
+    /// entire data stream before returning its stream of responses. The
+    /// recommended implementation of this function would start by reading the
+    /// message's content to a temporary file, and then produce a stream that
+    /// writes the message to all of the mailboxes one after the other.
+    ///
     /// Note: the EscapedDataReader has an inner buffer size of
     /// [`RDBUF_SIZE`](RDBUF_SIZE), which means that reads should not happen
     /// with more than this buffer size.
     ///
     /// Also, note that there is no timeout applied here, so the implementation
     /// of this function is responsible for making sure that the client does not
     /// just stop sending anything to DOS the system.
-    async fn handle_mail<'a, R>(
-        &self,
-        stream: &mut EscapedDataReader<'a, R>,
+    async fn handle_mail<'contents, 'cfg, 'connmeta, 'resp, R>(
+        &'cfg self,
+        stream: &mut EscapedDataReader<'contents, R>,
         meta: MailMetadata<Self::MailUserMeta>,
-        conn_meta: &mut ConnectionMetadata<Self::ConnectionUserMeta>,
-    ) -> Decision<()>
+        conn_meta: &'connmeta mut ConnectionMetadata<Self::ConnectionUserMeta>,
+    ) -> <Self::Protocol as Protocol<'resp>>::HandleMailReturnType
     where
-        R: Send + Unpin + AsyncRead;
+        R: Send + Unpin + AsyncRead,
+        'cfg: 'resp,
+        'connmeta: 'resp,
+        Self: 'resp;
```

Comment:
```suggestion
    async fn handle_mail<'borrowed, 'resp, R>(
        &'borrowed self,
        stream: &mut EscapedDataReader<'_, R>, // NOT borrowed
        meta: MailMetadata<Self::MailUserMeta>,
        conn_meta: &'borrowed mut ConnectionMetadata<Self::ConnectionUserMeta>,
    ) -> <Self::Protocol as Protocol<'resp>>::HandleMailReturnType
    where
        R: Send + Unpin + AsyncRead,
        'borrowed: 'resp,
        Self: 'resp;
```

I think it's possible to slightly simplify lifetime handling in this way. WDYT?